numerics/cbrt.cpp

```diff
@@ -40,42 +171,172 @@ double Cbrt(double const y) {
+  double const Δ = numerator / denominator;
+  double const r₀ = x_sign_y - Δ;
+  double const r₁ = x_sign_y - r₀ - Δ;
+
+  double const r̃ = r₀ + 2 * r₁;
+  if (rounding == Rounding::Correct &&
+      CorrectionPossiblyNeeded(r₀, r₁, r̃, /*τ=*/0x1.7C8587D10158Cp-13)) {
+    return _mm_cvtsd_f64(_mm_or_pd(
+        _mm_set_sd(CorrectLastBit(abs_y, std::abs(r₀), std::abs(r̃))), sign));
+  }
+  return r₀;
+}
+template double Cbrt<Rounding::Faithful>(double y);
+template double Cbrt<Rounding::Correct>(double y);
+}  // namespace method_3²ᴄZ5¹
+
+namespace method_5²Z4¹FMA {
+
+// No overflow or underflow occurs in intermediate computations for
+// y ∈ [y₁, y₂].  Steps 2 and 4 respectively involve the 6th powers of q and x,
+// approximately the 6th power of ∛y.  The 7th power of ∛y does not overflow for
+// y ∈ [y₁, y₂].  Here the range [y₁, y₂] is large enough that a single
+// rescaling suffices.
+constexpr double y₁ = 0x1p-438;
+constexpr double σ₁ = 0x1p-292;
+constexpr double σ₁⁻³ = 1 / (σ₁ * σ₁ * σ₁);
+constexpr double y₂ = 0x1p438;
+constexpr double σ₂ = 0x1p292;
+constexpr double σ₂⁻³ = 1 / (σ₂ * σ₂ * σ₂);
+static_assert(σ₁⁻³ * y₁ == y₂);
+static_assert(σ₂⁻³ * y₂ == y₁);
+static_assert(σ₁⁻³ * std::numeric_limits<double>::denorm_min() > y₁);
+static_assert(σ₂⁻³ * std::numeric_limits<double>::max() < y₂);
+
+template<Rounding rounding>
+double Cbrt(double const y) {
+  __m128d Y_0 = _mm_set_sd(y);
+  __m128d const sign = _mm_and_pd(masks::sign_bit, Y_0);
+  Y_0 = _mm_andnot_pd(masks::sign_bit, Y_0);
+  double const abs_y = _mm_cvtsd_f64(Y_0);
+
+  if (y != y) {
+    // The usual logic will produce a qNaN when given a NaN, but will not
+    // preserve the payload and will signal overflows (q will be a nonsensical
+    // large value, and q³ will overflow).  Further, the rescaling comparisons
+    // will signal the invalid operation exception for quiet NaNs (although that
+    // would be easy to work around using the unordered compare intrinsics).
+    return y + y;
+  }
+
+  if (abs_y < y₁) {
+    if (abs_y == 0) {
+      return y;
+    }
+    return Cbrt<rounding>(y * σ₁⁻³) * σ₁;
+  } else if (abs_y > y₂) {
+    if (abs_y == std::numeric_limits<double>::infinity()) {
+      return y;
+    }
+    return Cbrt<rounding>(y * σ₂⁻³) * σ₂;
+  }
+
+  // Step 1.  The constant Γᴋ minimizes the error of q, as in [KB01], rather
+  // than that of ξ.  This does not matter all that much here.
+  constexpr double Γᴋ = 0x0.19D9'06CB'2868'81F4'88FD'38DF'E7F6'98DD'Bp0;
+  constexpr std::uint64_t C = 0x2A9F'7625'3119'D328;
+  // C is defined as ⟦(2×1023 − Γᴋ) / 3⟧, which happens to differs by 296 ULPs
+  // from ⟦⟦2×1023 − ⟦Γᴋ⟧⟧ / 3⟧ which we compute in this static_assert.
+  static_assert((C - 296) * 0x1p-52 == (2 * 1023 - Γᴋ) / 3);
+  std::uint64_t const Y = _mm_cvtsi128_si64(_mm_castpd_si128(Y_0));
+  std::uint64_t const Q = C + Y / 3;
+  double const q = _mm_cvtsd_f64(_mm_castsi128_pd(_mm_cvtsi64_si128(Q)));
```

Comment:
There isn’t that much code there ; a lot of it is comments on the constants, which differ.
Factoring things out would be very messy since we have a lot of local variables that we want to use throughout.